homeassistant/components/sensor/__init__.py

```diff
@@ -127,3 +134,75 @@ def state_attributes(self) -> dict[str, Any] | None:
             return {ATTR_LAST_RESET: last_reset.isoformat()}
 
         return None
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the value reported by the sensor."""
+        return self._attr_native_value
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the sensor, if any."""
+        return self._attr_native_unit_of_measurement
+
+    @property
+    def unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the entity, after unit conversion."""
+        if self._attr_unit_of_measurement is not None:
+            return self._attr_unit_of_measurement
+
+        native_unit_of_measurement = self.native_unit_of_measurement
+
+        if native_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+            # Suppress AttributeError to handle tests testing directly on entity objects
+            with suppress(ValueError, AttributeError):
+                return self.hass.config.units.temperature_unit
+
+        return native_unit_of_measurement
+
+    @property
+    def state(self) -> Any:
+        """Return the state of the sensor and perform unit conversions, if needed."""
+        if self._attr_state is not None:
+            return self._attr_state
+
+        unit_of_measurement = self.native_unit_of_measurement
+        value = self.native_value
+
+        # Handle tests testing directly on entity objects without setting hass
+        if not self.hass:
+            return value
+
+        units = self.hass.config.units
+        if (
+            value is not None
+            and unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT)
+            and unit_of_measurement != units.temperature_unit
+        ):
+            if (
+                self.device_class != DEVICE_CLASS_TEMPERATURE
+                and not self._temperature_conversion_reported
+            ):
+                self._temperature_conversion_reported = True
+                report_issue = self._suggest_report_issue()
+                _LOGGER.warning(
+                    "Entity %s (%s) with device_class %s reports a temperature in "
+                    "%s which will be converted to %s, this is deprecated and will"
+                    " be removed from Home Assistant Core 2021.10. Please update "
+                    "your configuration if device_class is manually configured, "
+                    "otherwise %s",
+                    self.entity_id,
+                    type(self),
+                    self.device_class,
+                    unit_of_measurement,
+                    units.temperature_unit,
+                    report_issue,
+                )
+            value_s = STATE_UNKNOWN if value is None else str(value)
```

Comment:
Converting to `STATE_UNKNOWN` should happen in the base Entity class and not in the child classes. Additionally the lines below don't make sense if `value_s` is `"unknown"`.

Reply:
Fixed in bfbfa9a